A66-otel-stats.md

```diff
@@ -0,0 +1,395 @@
+# OpenTelemetry Metrics
+
+*   Author: Yash Tibrewal (@yashykt)
+*   Approver: Mark Roth (@markdroth)
+*   Status: In Review
+*   Implemented in: <language, ...>
+*   Last updated: Jul 20, 2023
+*   Discussion at: https://groups.google.com/g/grpc-io/c/po-deqYEQzE
+
+## Abstract
+
+Propose a metrics data model for gRPC OpenTelemetry metrics.
+
+## Background
+
+There are a collection of
+[metrics](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md)
+proposed by OpenCensus for gRPC. OpenCensus is no longer being actively
+maintained and is being
+[deprecated](https://opentelemetry.io/blog/2023/sunsetting-opencensus/#:~:text=Compatibility%20specification%204.-,What%20to%20Expect%20After%20July%2031st%2C%202023,found%20will%20not%20be%20patched.),
+with OpenTelemetry suggested as the successor framework.
+
+### Related Proposals:
+
+*   [A45: Exposing OpenCensus Metrics and Tracing for gRPC retry](A45-retry-stats.md)
+
+## Proposal
+
+### Metrics Schema
+
+#### Units
+
+Following the
+[OpenTelemetry Metrics Semantic Conventions](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/),
+the following units are used -
+
+*   Latencies are measured in float64 seconds, `s`
+*   Sizes are measured in bytes, `By`
+*   Counts for number of calls are measured in `{call}`
+*   Counts for number of attempts are measured in `{attempt}`
+
+Buckets for histograms in default views should be as follows -
+
+*   Latency : 0, 0.00001, 0.00005, 0.0001, 0.0003, 0.0006, 0.0008, 0.001, 0.002,
+    0.003, 0.004, 0.005, 0.006, 0.008, 0.01, 0.013, 0.016, 0.02, 0.025, 0.03,
+    0.04, 0.05, 0.065, 0.08, 0.1, 0.13, 0.16, 0.2, 0.25, 0.3, 0.4, 0.5, 0.65,
+    0.8, 1, 2, 5, 10, 20, 50, 100
+*   Size : 0, 1024, 2048, 4096, 16384, 65536, 262144, 1048576, 4194304,
+    16777216, 67108864, 268435456, 1073741824, 4294967296
+*   Count : 0, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192,
+    16384, 32768, 65536
+
+These buckets were chosen to maintain compatibility with the gRPC OpenCensus
+spec. The OpenTelemetry API has added an experimental feature for
+[advice](https://opentelemetry.io/docs/specs/otel/metrics/api/#instrument-advice)
+that would allow the gRPC library to provide these buckets as a hint. Since this
+is still an experimental feature and not yet implemented in all languages, it is
+upto the user to choose the right bucket boundaries.
+
+Also note that, as per an
+[OpenTelemetry proposal on stability](https://docs.google.com/document/d/1Nvcf1wio7nDUVcrXxVUN_f8MNmcs0OzVAZLvlth1lYY/edit#heading=h.dy1cg9doaq26)
+though, changes to bucket boundaries might not be considered a breaking change.
+Depending on the proposal, this recommendation would change to use
+`ExponentialHistogram`s instead, which would allow for automatic adjustments of
+the scale to better fit the data.
+
+#### Attributes
```

Comment:
Do you mean attributes (key-value pairs) that remain constant for the life of the plugin?

Reply:
Does `OTEL_RESOURCE_ATTRIBUTES` satisfy your usecase?

Reply:
I don't think it does. From what I understand using that environment variable is only useful for static global flags. What we're hoping for is something more dynamic. For example, some of our users like to specify a `source` tag that identifies a group of clients s.t. problems can be localized when they arise. Tags like these may differ within a given environment, so an environment variable is insufficient for setting those tags. Without the flexibility to add tags more dynamically in code, the metrics become rather inflexible and force us to separately define metrics when we otherwise would like to use the ones provided by grpc.

Reply:
Clarifying questions : Are you thinking of a case where these labels be the same within the whole process, or would different clients/servers in the same process also want different labels?

Reply:
The latter. The labels may be different for clients/servers within the same process. They may also be different across requests for the same client/server within a process (e.g. if we want to classify requests in some way). The more flexible it can be the better.

Reply:
This is the first I'm hearing about such a usecase, but that doesn't mean that we can't add it. Can you talk more about what kind of labels you are imagining? The source labels would not change for the life of the channel/server, so you should be able to add those at the point of configuring the plugin for the respective channel/server. As for the method, we already have the method name (and also the target name for clients).
((That would also help determine what information you would need for adding those labels.)

Reply:
Can that be added at the point of configuration? To my understanding source labels might change within the life of the server because they may differ per request. There are other per-request attributes that may be useful such as tags related to request content or authorization.

Reply:
This is tricky. For example, request content is not going to be available at the point of `started` because at that point you would only have the initial metadata. It won't be available at the point where the RPC ends either, since the RPC could potentially have been a streaming RPC.

I think we might be able to provide an API to inject labels based on the initial metadata. The issue with that currently is that atleast in gRPC Core, the types used in stats plugins like OTel plugins are not stable, and they are still going through enough change. This might be possible in other language implementations though.

Reply:
@markdroth @ejona86 @dfawley what do you think?

Reply:
It's not clear to me whether the request here is to allow setting labels based on headers or based on message payloads.

I am pretty strongly opposed to the idea of allowing labels to be set based on message payloads, for the following reasons:
- gRPC at its core treats payloads as opaque blobs and does not make assumptions about what format the payloads are in.  While most users use gRPC with protobuf payloads, gRPC can be used with other serialization formats, and we don't want to break that by baking in assumptions that the payloads are in protobuf form.
- There are a bunch of layering problems here.  At the layer at which gRPC core is interacting with the CallTracer, the payloads may not be represented in a form suitable to export in any sort of public API.  And the payloads are typically just serialized bytes, so even if they were in protobuf form, it would add quite a bit of overhead to deserialize them just to figure out what label to add.
- As Yash mentioned, the entire idea of setting labels based on message contents inherently seems to assume that the requests are unary (i.e., there is only one message payload per RPC).  However, in gRPC, streaming is a first-class use-case (in fact, unary is actually implemented as a special case of streaming), and we don't want to build features that work only for unary.

Using headers instead of payloads is less objectionable to me than using message payloads, but still not a slam dunk.  Here are some factors we'd need to consider:
- My main concern is that if we give users the ability to add arbitrary labels based on the headers or message payloads, it will be very hard to prevent them from hitting cardinality limits for the resulting timeseries.  This can be a problem both in OT storing the data in memory and in whatever monitoring system you eventually export the OT data to.  It's not clear to me how we can provide this flexibility without making it too easy for users to shoot themselves in the foot.
- Even for headers, there can be significant overhead here, especially if we would need to support things like regex matches.  I'm concerned that we would again be giving users the ability to shoot themselves in the foot performance-wise.
- There are some API/layering concerns here similar to those that I mentioned for message payloads, although they are somewhat more tractable for headers than they are for message payloads.  We might be able to find a way to make this workable, but it will take some careful thought.

Reply:
How about we do this:
* allow setting labels only based on headers.
* require the header to have full key=value pair definition, no regexes and no additional parsing allowed.

This still should cover the main use-case  we have in mind for this. If we need to group clients on a server by client type, we'll simply add a header "source=clientA" via an interceptor on every client.  Then on the server side we will be able to filter requests by client type, which is super useful IMO.  In our case client type is the name of the client service, it is always static and it has very low cardinality, but I agree that people may misuse this feature and put some very dynamic data in it. 

I think this addresses some of the Mark's concerns:
* the overhead should be minimal, as we are doing no additional processing and parsing headers is cheap and should be done early in the call stack anyway.
* It is harder to accidentally get explosive cardinality if you are forced to put the exact values in the header. We can also provide a hardcoded limit and completely remove a tag on the server after its cardinality exceeds the threshold. In this case users will either notice a warn message in the server logs or notice the the tag is gone, so they could either fix the cardinality of their tag or bump the threshold via some configurable parameter.

Reply:
I'm not really keen on having an API that explicitly provides the functionality that allows header values to dictate metric label values directly.  I think that makes it a little too easy for users to shoot themselves in the foot in terms of cardinality.  I think I'd rather provide an API hook that calls some user-supplied code to read the headers and return the set of labels to add, which would need to return the result synchronously.  That does still give users the ability to shoot themselves in the foot, but at least they have to write a bit of code to do so, so at least we won't be directly encouraging dangerous functionality.

I think this probably needs a bit more consideration before we decide.  Let's keep talking about this separately, but we need to get the basic OT metric functionality nailed down fairly soon, so I think we'll exclude this from this gRFC for now.  However, there will definitely be more gRFCs adding more OT metrics functionality on top of the foundation laid by this gRFC, so we'll definitely have opportunities to add this once we have consensus on it.

Reply:
As discussed, I've stripped the proposal to the bare minimal stuff that we all agree on.